### ISSUE TITLE
Add cosine annealing for learning rate decay

### ISSUE THREAD

**Issue by adler-j** [tensorflow/tensorflow#11113]
[SGDR: Stochastic Gradient Descent With Warm Restarts](https://openreview.net/pdf?id=Skq89Scxx), proposes decaying the learning rate according to

![image](https://user-images.githubusercontent.com/2202312/27641761-1bf302c8-5c1d-11e7-8d4b-15988701ff3f.png)

where ![image](https://user-images.githubusercontent.com/2202312/27641775-28b4e0a8-5c1d-11e7-8adc-1dcaea55d77c.png) is the minimum step length, ![image](https://user-images.githubusercontent.com/2202312/27641798-3be6a80a-5c1d-11e7-9b41-d3c25b0b2b96.png) is the maximum step length, ![image](https://user-images.githubusercontent.com/2202312/27641818-4686daf0-5c1d-11e7-8991-533ce8710e8f.png) is the global step and ![image](https://user-images.githubusercontent.com/2202312/27641841-5594f2ac-5c1d-11e7-9f54-e3cc8ccd1566.png) is the maximum number of iterations.

I've personally found this strategy to  be easy to use given that the number of hyperparameters is relatively small and results are good.

Is this something we want added to tensorflow? Would you accept submissions?

**Comment by shin-migami** [comment]
Can i get assigned to this ?

**Comment by skye** [comment]
@ebrevdo @martinwicke can you comment?

**Comment by martinwicke** [comment]
@shin-migami Consider yourself assigned!

The right spot for this is in contrib.

I cannot formally assign you, but everybody who wants to work on this should check this issue first, and find that you are already working on it. Please reference this issue in any PR you create, and update the issue when you are making progress.

**Comment by michaelisard** [comment]
Adding contributions welcome tag just so it's clear that this issue has been triaged: this doesn't contradict the fact that @shin-migami is working on it!

**Comment by PatrykChrabaszcz** [comment]
Hi @shin-migami 
I was asked by Ilya and Frank (authors of the paper) to implement this learning rate decay algorithm in tensorflow. Now I found out you also work on it. 

We would like to include more general version where you can also use restarts.
Did you make any progress already?

**Comment by adler-j** [comment]
I guess this can be closed now?
